docs/proposals/compute-resource-metrics-api.md

```diff
@@ -203,63 +170,99 @@ type DerivedMetrics struct {
-(e.g. last 1s) or updated more often, they should use the metrics pipeline via
-Heapster.
-
-The API server will not hold any of this data directly. For our initial
-purposes, it will hold the most recent metrics obtained from each node in
-etcd. Then, when polled for metrics, the API server would only serve said most
-recent data per node.
-
-Benchmarks will be run with etcd to see if it can keep up with the frequent
-writes of data. If it turns out that etcd doesn't scale well enough, we will
-have to switch to a different storage system.
+API server will have the functionality for Metrics API built into it.
+Clients will be able to Get and Watch metrics for nodes and pods.
+The API server will hold only the most recent Metrics sample.
 
 If a pod gets deleted, the API server will get rid of any metrics it may
 currently be holding for it.
 
 The clients watching the metrics data may cache it for longer periods of time.
-The clearest example would be Heapster.
 
-In the future, we might want to store the metrics differently:
+The primary issue with the approach is that etcd is not built to handle metrics data.
+Assuming minutely resolution, 100 nodes and 100 pods per node, the API server will have to handle ~168 QPS.
+Metrics don't have to be persisted.
+Old metrics are not of much use.
+Hence we can avoid storing to etcd and instead store in memory or to an alternate storage.
+If the API server is distributed, the storage medium has to be chosen appropriately.
+
+The source of the Metrics data can either be kubelet or heapster as of now.
+
+##### Kubelet as the source
+
+Kubelet has cAdvisor compiled into it.
+Hence it has access to all the required metrics.
+It is already built to be able to write to the API server.
+Computing these metrics in the kubelet will help with scalability of the cluster, when compared to computing at the cluster level.
+
+The primary issue with this approach is that of metrics persistence.
+Kubelet has [no checkpointing](https://issues.k8s.io/489) functionlity as of now.
+Kubelet stores metrics data in memory as of now.
+Kubelet restarts will wipe out all historical metrics.
+In-order to survive restarts, metrics data will have to be persisted either directly through checkpoints or by storing data to a local database.
+TODO(vishh@): Explore possible options for local databases.
+
+##### Heapster as the source
+
+Heapster queries metrics from all the nodes and stores them to a configurable backend.
+Heapster computes statistical data based on the metrics (cpu, memory & filesystem) and caches the statistical data for upto a day.
+It exposes the same data via REST endpoints.
+Metrics data can be either pushed from heapster to the API server either directly or using a helper process.
+Since heapster has access to permanent storage, it can re-build its internal cache after restarts,
+
+This approach makes heapster a core component.
+Heapster will have be run on every kubernetes cluster.
+
+
+#### Third-party API server
```

Comment:
I'm not sure "third party" is the right name for this. I think we all agreed that these core metrics need to be first-party, even if they're not served by the core API server.